Expand documentation on setup hooks (#1023)
software-mansion/protostar: website/docs/tutorials/07-testing/README.md

The setup case hook is bound to a matching test case and is executed just before the test case
itself.
These hooks are executed with a context built by the `__setup__` hook, but in isolation for each

> test separately, before jumping into test case function itself.

Reviewer: - It's not obvious what `these` refers to, because the previous sentence uses the singular form.
- This sentence could be split in two.

Author: 6f5132f

This makes them useful to extract test-specific setup logic from tested code itself.

```cairo title="Using setup case hook to prepare test-specific state"
@external
func __setup__{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {
    balance.write(10);
    return ();
}

@external
func setup_need_more_money{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {
    balance.write(10000);
    return ();
}

@external
func test_need_more_money{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {
    alloc_locals;
    let (amount_ref) = balance.read();
    local amount = amount_ref;

    assert amount = 10000;

    return ();
}

@external
func test_foo{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {
    alloc_locals;
    let (amount_ref) = balance.read();
    local amount = amount_ref;

    assert amount = 10;

    return ();
}
```

> You can also use case hooks to configure the behavior of a particular test case,

Reviewer: ```suggestion
You can also use setup case hooks to configure the behavior of a particular test case,
```

The setup case hook is bound to a matching test case and is executed just before the test case
itself.

> Setup case hook is executed with a context built by the `__setup__` hook,

Reviewer: ```suggestion
The hook is executed within a context built by the `__setup__` hook,
```

> but in isolation for each test separately.

Author: Done

> After it, Protostar immediately executes the test case function itself.

Reviewer: ```suggestion
Then, Protostar immediately executes the test case function.
```

This makes them useful to extract test-specific setup logic from tested code itself.

```cairo title="Using setup case hook to prepare test-specific state"
@external
func __setup__{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {
    balance.write(10);
    return ();
}

@external
func setup_need_more_money{
    syscall_ptr: felt*,
    pedersen_ptr: HashBuiltin*,
    range_check_ptr
}() {

>     balance.write(10000);

Reviewer 2: Should we add balance definition to the code? Somebody may get confused

Author: I have thought about doing so, but decided not to because this example grew to enormous size and is occupying a lot of eye space. In the same manner, I have removed imports and `%lang starknet` lines from example.

Reviewer 2: We will add it if somebody asks about it
